### PR TITLE
Draft: [GPU] L2 prefetch.

### PR DESCRIPTION
**Pull request by sergachev**

### CODE REVIEW

**Comment by sergachev** [comment]
Example use:

```
import jax
from jax.experimental.xla_metadata import set_xla_metadata
import jax.numpy as jnp
from jax.sharding import PartitionSpec as P


mesh = jax.make_mesh((2, 4), ("x", "y"))
sharding = jax.sharding.NamedSharding(mesh, P("x", "y"))
arr = jnp.arange(32000.0).reshape(40, 800)
arr_sharded = jax.device_put(arr, sharding)


def f(x):
    with set_xla_metadata(l2_prefetch_opportunity=10000, prefetch_num_blocks=4):
        # becomes all-reduce sum
        return x.sum(axis=0)


def g(x, y):
    return f(x) * y


lowered = jax.jit(g).lower(arr_sharded, arr_sharded)
compiled_hlo = lowered.compile().as_text()
assert 'custom_call_target="l2_prefetch"' in compiled_hlo
```

**Comment by sergachev** [comment]
One possible microbenchmark:

```
HloModule m, is_scheduled=true

gemm_fusion0 {
 a = bf16[3,1024,1024] parameter(0)
 b = bf16[2048,1024] parameter(1)
 s = bf16[1,1024] slice(b), slice={[0:1],[0:1024]}
 r = bf16[1024] bitcast(s)
 c = bf16[3,1024,1024] broadcast(r), dimensions={2}
 d = bf16[3,1024,1024] dot(c, a),
    lhs_batch_dims={0}, lhs_contracting_dims={2},
    rhs_batch_dims={0}, rhs_contracting_dims={1}
}

gemm_fusion1 {
 a = bf16[3,1024,1024] parameter(0)
 b = bf16[2048,1024] parameter(1)
 s = bf16[1,1024] slice(b), slice={[0:1],[0:1024]}
 r = bf16[1024] bitcast(s)
 c = bf16[3,1024,1024] broadcast(r), dimensions={2}
 d = bf16[3,1024,1024] dot(c, a),
    lhs_batch_dims={0}, lhs_contracting_dims={2},
    rhs_batch_dims={0}, rhs_contracting_dims={1}
}

flush {
  p = bf16[30,1024,1024] parameter(0)
  n = bf16[30,1024,1024] negate(p)
}

entry {
 p0 = bf16[256,1024] parameter(0)
 p1 = bf16[3,1024,1024] parameter(1)
 p2 = bf16[3,1024,1024] parameter(2)
 p3 = bf16[30,1024,1024] parameter(3)

 p0c = bf16[256,1024] copy(p0)

 flush = bf16[30,1024,1024] fusion(p3), kind=kLoop, calls=flush

 ag-start = (bf16[256,1024], bf16[2048,1024]) all-gather-start(p0c),
  dimensions={0}, replica_groups={}, control-predecessors={flush}
 ag-done = bf16[2048,1024] all-gather-done(ag-start),
  frontend_attributes={l2_prefetch_opportunity="16000000",
                       prefetch_num_blocks="4"}

 g0 = bf16[3,1024,1024] fusion(p1, ag-done), kind=kCustom, calls=gemm_fusion0,
  backend_config={"fusion_backend_config":{"kind":"__triton_gemm",
  "triton_gemm_config":
  {"block_m":"128","block_n":"128","block_k":"32","split_k":"1",
   "num_stages":"5","num_warps":"8","num_ctas":"1"}}}
 g1 = bf16[3,1024,1024] fusion(p2, ag-done), kind=kCustom, calls=gemm_fusion1,
  backend_config={"fusion_backend_config":{"kind":"__triton_gemm",
  "triton_gemm_config":
  {"block_m":"128","block_n":"128","block_k":"32","split_k":"1",
   "num_stages":"5","num_warps":"8","num_ctas":"1"}}}, control-predecessors={g0}
}
```

```
bazel run xla/tools/multihost_hlo_runner/hlo_runner_main -- --profile_execution --run_xla_backend_only --hlo_argument_mode=uninitialized --output_mode=not_return_outputs --num_repeats=100 --xla_gpu_graph_min_graph_size=1 --xla_gpu_enable_command_buffer=fusion,collectives --xla_gpu_graph_enable_concurrent_region x.hlo

## Execution time, file=/w/1.hlo repeat=90 duration=127456ns
## Execution time, file=/w/1.hlo repeat=91 duration=127456ns
## Execution time, file=/w/1.hlo repeat=92 duration=129760ns
## Execution time, file=/w/1.hlo repeat=93 duration=127936ns
## Execution time, file=/w/1.hlo repeat=94 duration=128576ns
## Execution time, file=/w/1.hlo repeat=95 duration=129056ns
## Execution time, file=/w/1.hlo repeat=96 duration=129312ns
## Execution time, file=/w/1.hlo repeat=97 duration=127872ns
## Execution time, file=/w/1.hlo repeat=98 duration=126240ns
## Execution time, file=/w/1.hlo repeat=99 duration=126304ns
```

Same with `--xla_disable_hlo_passes=l2-prefetch-scheduler` (reference measurement):
```
## Execution time, file=/w/1.hlo repeat=90 duration=130496ns
## Execution time, file=/w/1.hlo repeat=91 duration=129120ns
## Execution time, file=/w/1.hlo repeat=92 duration=130816ns
## Execution time, file=/w/1.hlo repeat=93 duration=131072ns
## Execution time, file=/w/1.hlo repeat=94 duration=129824ns
## Execution time, file=/w/1.hlo repeat=95 duration=131456ns
## Execution time, file=/w/1.hlo repeat=96 duration=131360ns
## Execution time, file=/w/1.hlo repeat=97 duration=130720ns
## Execution time, file=/w/1.hlo repeat=98 duration=131200ns
## Execution time, file=/w/1.hlo repeat=99 duration=130112ns
```

**Comment by sergachev** [comment]
Another microbenchmark, runs with the same commands as the previous one.

```
## Execution time, file=/w/1.hlo repeat=90 duration=312480ns
## Execution time, file=/w/1.hlo repeat=91 duration=330144ns
## Execution time, file=/w/1.hlo repeat=92 duration=313632ns
## Execution time, file=/w/1.hlo repeat=93 duration=321888ns
## Execution time, file=/w/1.hlo repeat=94 duration=310976ns
## Execution time, file=/w/1.hlo repeat=95 duration=322176ns
## Execution time, file=/w/1.hlo repeat=96 duration=318976ns
## Execution time, file=/w/1.hlo repeat=97 duration=313248ns
## Execution time, file=/w/1.hlo repeat=98 duration=325568ns
## Execution time, file=/w/1.hlo repeat=99 duration=315520ns
```
->
```
## Execution time, file=/w/1.hlo repeat=90 duration=281184ns
## Execution time, file=/w/1.hlo repeat=91 duration=278112ns
## Execution time, file=/w/1.hlo repeat=92 duration=278016ns
## Execution time, file=/w/1.hlo repeat=93 duration=278528ns
## Execution time, file=/w/1.hlo repeat=94 duration=288192ns
## Execution time, file=/w/1.hlo repeat=95 duration=277920ns
## Execution time, file=/w/1.hlo repeat=96 duration=293440ns
## Execution time, file=/w/1.hlo repeat=97 duration=280960ns
## Execution time, file=/w/1.hlo repeat=98 duration=284032ns
## Execution time, file=/w/1.hlo repeat=99 duration=269216ns
```

```
HloModule m, is_scheduled=true

flush {
  p = bf16[50,1024,1024] parameter(0)
  ROOT n = bf16[50,1024,1024] negate(p)
}

gemm_1_0 {
  b.16 = bf16[2048,1024] parameter(1)
  s.16 = bf16[1,1024] slice(b.16), slice={[0:1], [0:1024]}
  r.16 = bf16[1024] bitcast(s.16)
  c.16 = bf16[1,512,1024] broadcast(r.16), dimensions={2}
  a.16 = bf16[1,512,1024] parameter(0)
  ROOT d.16 = bf16[1,512,512] dot(c.16, a.16), lhs_batch_dims={0}, lhs_contracting_dims={2}, rhs_batch_dims={0}, rhs_contracting_dims={2}
}

gemm_1_1 {
  b.17 = bf16[2048,1024] parameter(1)
  s.17 = bf16[1,1024] slice(b.17), slice={[0:1], [0:1024]}
  r.17 = bf16[1024] bitcast(s.17)
  c.17 = bf16[1,512,1024] broadcast(r.17), dimensions={2}
  a.17 = bf16[1,512,1024] parameter(0)
  ROOT d.17 = bf16[1,512,512] dot(c.17, a.17), lhs_batch_dims={0}, lhs_contracting_dims={2}, rhs_batch_dims={0}, rhs_contracting_dims={2}
}

gemm_1_2 {
  b.18 = bf16[2048,1024] parameter(1)
  s.18 = bf16[1,1024] slice(b.18), slice={[0:1], [0:1024]}
  r.18 = bf16[1024] bitcast(s.18)
  c.18 = bf16[1,512,1024] broadcast(r.18), dimensions={2}
  a.18 = bf16[1,512,1024] parameter(0)
  ROOT d.18 = bf16[1,512,512] dot(c.18, a.18), lhs_batch_dims={0}, lhs_contracting_dims={2}, rhs_batch_dims={0}, rhs_contracting_dims={2}
}

gemm_1_3 {
  b.19 = bf16[2048,1024] parameter(1)
  s.19 = bf16[1,1024] slice(b.19), slice={[0:1], [0:1024]}
  r.19 = bf16[1024] bitcast(s.19)
  c.19 = bf16[1,512,1024] broadcast(r.19), dimensions={2}
  a.19 = bf16[1,512,1024] parameter(0)
  ROOT d.19 = bf16[1,512,512] dot(c.19, a.19), lhs_batch_dims={0}, lhs_contracting_dims={2}, rhs_batch_dims={0}, rhs_contracting_dims={2}
}

gemm_1_4 {
  b.20 = bf16[2048,1024] parameter(1)
  s.20 = bf16[1,1024] slice(b.20), slice={[0:1], [0:1024]}
  r.20 = bf16[1024] bitcast(s.20)
  c.20 = bf16[1,512,1024] broadcast(r.20), dimensions={2}
  a.20 = bf16[1,512,1024] parameter(0)
  ROOT d.20 = bf16[1,512,512] dot(c.20, a.20), lhs_batch_dims={0}, lhs_contracting_dims={2}, rhs_batch_dims={0}, rhs_contracting_dims={2}
}

gemm_1_5 {
  b.21 = bf16[2048,1024] parameter(1)
  s.21 = bf16[1,1024] slice(b.21), slice={[0:1], [0:1024]}
  r.21 = bf16[1024] bitcast(s.21)
  c.21 = bf16[1,512,1024] broadcast(r.21), dimensions={2}
  a.21 = bf16[1,512,1024] parameter(0)
  ROOT d.21 = bf16[1,512,512] dot(c.21, a.21), lhs_batch_dims={0}, lhs_contracting_dims={2}, rhs_batch_dims={0}, rhs_contracting_dims={2}
}

gemm_1_6 {
  b.22 = bf16[2048,1024] parameter(1)
  s.22 = bf16[1,1024] slice(b.22), slice={[0:1], [0:1024]}
  r.22 = bf16[1024] bitcast(s.22)
  c.22 = bf16[1,512,1024] broadcast(r.22), dimensions={2}
  a.22 = bf16[1,512,1024] parameter(0)
  ROOT d.22 = bf16[1,512,512] dot(c.22, a.22), lhs_batch_dims={0}, lhs_contracting_dims={2}, rhs_batch_dims={0}, rhs_contracting_dims={2}
}

gemm_1_7 {
  b.23 = bf16[2048,1024] parameter(1)
  s.23 = bf16[1,1024] slice(b.23), slice={[0:1], [0:1024]}
  r.23 = bf16[1024] bitcast(s.23)
  c.23 = bf16[1,512,1024] broadcast(r.23), dimensions={2}
  a.23 = bf16[1,512,1024] parameter(0)
  ROOT d.23 = bf16[1,512,512] dot(c.23, a.23), lhs_batch_dims={0}, lhs_contracting_dims={2}, rhs_batch_dims={0}, rhs_contracting_dims={2}
}

ENTRY entry {
  p0 = bf16[256,1024] parameter(0)
  p1 = bf16[50,1024,1024] parameter(1)
  p2 = bf16[1,512,1024] parameter(2)
  p3 = bf16[1,512,1024] parameter(3)
  p4 = bf16[1,512,1024] parameter(4)
  p5 = bf16[1,512,1024] parameter(5)
  p6 = bf16[1,512,1024] parameter(6)
  p7 = bf16[1,512,1024] parameter(7)
  p8 = bf16[1,512,1024] parameter(8)
  p9 = bf16[1,512,1024] parameter(9)

  p0c1 = bf16[256,1024] copy(p0)
  flush = bf16[50,1024,1024] fusion(p1), kind=kLoop, calls=flush
  ag-start1 = (bf16[256,1024], bf16[2048,1024]) all-gather-start(p0c1), replica_groups={}, dimensions={0}
  ag-done1 = bf16[2048,1024] all-gather-done(ag-start1), frontend_attributes={l2_prefetch_opportunity="8388608",prefetch_num_blocks="4"}

  g_1_0 = bf16[1,512,512] fusion(p2, ag-done1), kind=kCustom, calls=gemm_1_0, backend_config={"fusion_backend_config":{"kind":"__triton_gemm",
   "triton_gemm_config":
   {"block_m":"128","block_n":"128","block_k":"32","split_k":"1",
    "num_stages":"5","num_warps":"8","num_ctas":"1"}}}
  g_1_1 = bf16[1,512,512] fusion(p3, ag-done1), kind=kCustom, calls=gemm_1_1, backend_config={"fusion_backend_config":{"kind":"__triton_gemm",
   "triton_gemm_config":
   {"block_m":"128","block_n":"128","block_k":"32","split_k":"1",
    "num_stages":"5","num_warps":"8","num_ctas":"1"}}}, control-predecessors={g_1_0}
  g_1_2 = bf16[1,512,512] fusion(p4, ag-done1), kind=kCustom, calls=gemm_1_2, backend_config={"fusion_backend_config":{"kind":"__triton_gemm",
   "triton_gemm_config":
   {"block_m":"128","block_n":"128","block_k":"32","split_k":"1",
    "num_stages":"5","num_warps":"8","num_ctas":"1"}}}, control-predecessors={g_1_1}
  g_1_3 = bf16[1,512,512] fusion(p5, ag-done1), kind=kCustom, calls=gemm_1_3, backend_config={"fusion_backend_config":{"kind":"__triton_gemm",
   "triton_gemm_config":
   {"block_m":"128","block_n":"128","block_k":"32","split_k":"1",
    "num_stages":"5","num_warps":"8","num_ctas":"1"}}}, control-predecessors={g_1_2}
  g_1_4 = bf16[1,512,512] fusion(p6, ag-done1), kind=kCustom, calls=gemm_1_4, backend_config={"fusion_backend_config":{"kind":"__triton_gemm",
   "triton_gemm_config":
   {"block_m":"128","block_n":"128","block_k":"32","split_k":"1",
    "num_stages":"5","num_warps":"8","num_ctas":"1"}}}, control-predecessors={g_1_3}
  g_1_5 = bf16[1,512,512] fusion(p7, ag-done1), kind=kCustom, calls=gemm_1_5, backend_config={"fusion_backend_config":{"kind":"__triton_gemm",
   "triton_gemm_config":
   {"block_m":"128","block_n":"128","block_k":"32","split_k":"1",
    "num_stages":"5","num_warps":"8","num_ctas":"1"}}}, control-predecessors={g_1_4}
  g_1_6 = bf16[1,512,512] fusion(p8, ag-done1), kind=kCustom, calls=gemm_1_6, backend_config={"fusion_backend_config":{"kind":"__triton_gemm",
   "triton_gemm_config":
   {"block_m":"128","block_n":"128","block_k":"32","split_k":"1",
    "num_stages":"5","num_warps":"8","num_ctas":"1"}}}, control-predecessors={g_1_5}
  g_1_7 = bf16[1,512,512] fusion(p9, ag-done1), kind=kCustom, calls=gemm_1_7, backend_config={"fusion_backend_config":{"kind":"__triton_gemm",
   "triton_gemm_config":
   {"block_m":"128","block_n":"128","block_k":"32","split_k":"1",
    "num_stages":"5","num_warps":"8","num_ctas":"1"}}}, control-predecessors={g_1_6}
  
  t = tuple(flush, g_1_0, g_1_1, g_1_2, g_1_3, g_1_4, g_1_5, g_1_6, g_1_7)
}
```
